### PR TITLE
migrate ros2 devel

### PR DESCRIPTION
**Pull request by Karsten1987**
Same thing as https://github.com/ros-drivers/urg_c/pull/11 for the `urg_node`.

### CODE REVIEW

**Comment by Karsten1987** [comment]
I've added some touchups on top of rebasing the existing ros2 PR. I would leave it for as now as such, as I believe this PR is already hard enough to review.

**Comment by tonybaltovski** [comment]
@Karsten1987 will do, I'll also give it a try once I get access to some hardware.

**Comment by Karsten1987** [comment]
enabling the linters on this repo would also result in a huge change. So I'll enable `pedantic` to cover some basic rules and warning levels.

**Comment by airballking** [comment]
@Karsten1987 Great that you're looking into this! 

I also want to use ```urg_node``` on ROS2. So, I gave your fork a try and ran into a compilation issue that I could not fix (and two that I could fix). Are you interested in a report, i.e. shall I open an issue on your fork?

Bold question, I know: Do you have a timeline for a release of ```urg_node```?

**Comment by Karsten1987** [comment]
@airballking please just comment straight up here on this PR with the problems you encounter. This makes the most sense so we can directly iterate over this before merging.

as for releasing, I believe we should get things merged first and then it's essentially up to @tonybaltovski to decide when to release.

**Comment by Karsten1987** [comment]
If I think about it a bit more, I believe it makes sense to have the URG node developed as a lifecycle node. That would give the user a bit more flexibility to start/stop the scanning as well as diagnostics, but especially allows to have a clean reconfigure step.

Opinions?

**Comment by airballking** [comment]
I agree that having the node with lifecycle management is preferable. I think this would be important for an important HW device, such as the laser scanner.

**Comment by airballking** [comment]
@Karsten1987 thanks for the quick response! Then, I will double-check the things that I found and report them here.

**Comment by airballking** [comment]
@Karsten1987 Finally, I am back at getting ```urg_node``` to run! I think my issue is that I use the wrong forks/branches of one or several of the dependencies of the package. Could you please verify that I use the right things for compiling?

ROS2 version: Dashing
Installed: From debs
OS: Ubuntu 18.04

Fresh workspace with:
* diagnostics (https://github.com/ros/diagnostics.git, branch: ros2-devel)
* laser_proc (https://github.com/bosch-robotics-na/laser_proc.git, branch: ros2-devel)
* urg_c (https://github.com/bosch-robotics-na/urg_c.git, branch: ros2-devel)
* urg_node_msgs (https://github.com/bponsler/urg_node_msgs.git, branch: master)
* urg_node (https://github.com/bosch-robotics-na/urg_node.git, branch: ros2-devel)

If I call ```colcon build``` on this, then I get the following error:
```shell
Starting >>> urg_node
--- stderr: urg_node                                                                       

/home/georg/urg/src/urg_node/src/urg_node_driver.cpp: In member function ‘void urg_node::UrgNode::updateDiagnostics()’:
/home/georg/urg/src/urg_node/src/urg_node_driver.cpp:410:32: error: ‘void diagnostic_updater::Updater::update()’ is private within this context
     diagnostic_updater_.update();
                                ^
In file included from /home/georg/urg/src/urg_node/include/urg_node/urg_node_driver.hpp:43:0,
                 from /home/georg/urg/src/urg_node/src/urg_node_driver.cpp:34:
/home/georg/urg/src/install/diagnostic_updater/include/diagnostic_updater/diagnostic_updater.hpp:500:8: note: declared private here
   void update()
        ^~~~~~
make[2]: *** [CMakeFiles/urg_node_driver.dir/src/urg_node_driver.cpp.o] Error 1
make[1]: *** [CMakeFiles/urg_node_driver.dir/all] Error 2
make: *** [all] Error 2
---
Failed   <<< urg_node	[ Exited with code 2 ]

```

Seems like the APIs don't match. Hence, my guess about a wrong checkout...

**Comment by airballking** [comment]
I went and made the brutal thing: I commented all occurrences of ```diagnostics``` in ```urg_node```. With that, I got it to compile. ;)

When trying to run the node on our scanner (UAM-05LP-T301) I ran into some run-time issues:
- The node does not attempt to connect to the device.
- The acquisition thread is not started.
- The parameters are not read from the server.

To get some data fast, I fixed all of those for our device _in a non-generic fashion_ (aka hacks). With that, it works. In case someone is interested, you can find my fork [here](https://github.com/airballking/urg_node/tree/ros2-devel).

@Karsten1987 If you need someone to test a later version on real hardware, please don't hesitate to contact me. I am willing to help. We have further Hokuyo types in our lab...

**Comment by Paul-Ajay** [comment]
I got the same error as @airballking described. Looks like some API mismatch. Changing the `diagnostic_updater_.update();` to `diagnostic_updater_.force_update();` resolved the compilation issue.
But I am not able to see any sensor data on the /scan topic. But I am able to ping the sensor IP and the `ros2 run urg_node urg_node __params:=src/urg_pkgs/urg_node/launch/urg_node_ip.yaml` command is not giving any errors or warnings
I have the Hokuyo UST10LX lidar and it works properly with ROS Melodic
What am I missing?

File: src/urg_node.cpp
Comment (Paul-Ajay) [inline src/urg_node.cpp]: Accidently commented ? I couldn't find run() is not getting called anywhere else

File: src/urg_node_driver.cpp
Comment (Paul-Ajay) [inline src/urg_node_driver.cpp]: force_update()? I am getting compilation errors saying update() is private function

File: src/urg_node_driver.cpp
Comment (Paul-Ajay) [inline src/urg_node_driver.cpp]: Accidentally commented?

**Comment by Paul-Ajay** [comment]
Got it working after hard-coding the IP address and doing the above mentioned changes. Looks like the parameters are not getting added to the urg_node_driver

**Comment by Karsten1987** [comment]
Sorry y'all. This wasn't on my radar for the last couple of weeks. 

@ajaypaul2008 sounds great that you've got it working. Do you mind opening a PR against this PR - against https://github.com/bosch-robotics-na/urg_node/tree/ros2-devel?

Regarding the parameters: ROS2 dashing still requires the parameters being set via a `*.yaml` file. Starting from eloquent, individual parameters can conveniently be set via the command line in the style of `--ros-args -p <key>:=<value>`. More details here: https://index.ros.org/doc/ros2/Tutorials/Node-arguments/#parameters

File: src/urg_node_driver.cpp
Comment (Karsten1987) [reply src/urg_node_driver.cpp]: Most certainly. I could see myself doing that while trying to make this compile without having a physical device at hand.

File: src/urg_node_driver.cpp
Comment (Karsten1987) [reply src/urg_node_driver.cpp]: Yup. See here: https://github.com/ros/diagnostics/pull/114/files#r320031172

File: src/urg_node.cpp
Comment (Karsten1987) [reply src/urg_node.cpp]: Same comment as below. Most likely by accident.

File: CMakeLists.txt
Comment (clalancette) [inline CMakeLists.txt]: I'd just remove all references to `dynamic_reconfigure` everywhere in this PR; it is fully replaced by parameters in ROS 2, and won't likely be coming back.

File: CMakeLists.txt
Comment (clalancette) [inline CMakeLists.txt]: As far as I can tell, we are no longer using `boost` at all in this PR.  If that is the case, please remove this (and any references in the package.xml).

File: README.md
Comment (clalancette) [inline README.md]:
```suggestion
ROS 2 wrapper for the Hokuyo urg_c library.
```

File: README.md
Comment (clalancette) [inline README.md]:
```suggestion
Install ROS 2 as per instructions [here](https://index.ros.org/doc/ros2/Installation/#installationguide)
```

File: README.md
Comment (clalancette) [inline README.md]:
```suggestion
More information about colcon is available [here](https://media.readthedocs.org/pdf/colcon/latest/colcon.pdf)
```

File: README.md
Comment (clalancette) [inline README.md]:
```suggestion
A urdf file is already included and if you have succesfully ran an `colcon build`, the path to urdf file will be :
```

File: include/urg_node/urg_node_driver.hpp
Comment (clalancette) [inline include/urg_node/urg_node_driver.hpp]: Since we've renamed the file, also rename the header guards:

```
#ifndef URG_NODE_URG_NODE_DRIVER_HPP
#define URG_NODE_URG_NODE_DRIVER_HPP
```

File: include/urg_node/urg_node_driver.hpp
Comment (clalancette) [inline include/urg_node/urg_node_driver.hpp]: If you have this also take a `const rclcpp::NodeOptions & options` parameter, it probably isn't too much further work to make this composable.

File: launch/urg_node.yaml
Comment (clalancette) [inline launch/urg_node.yaml]: Nit: add trailing newline.

File: launch/urg_node_ip.yaml
Comment (clalancette) [inline launch/urg_node_ip.yaml]: Nit: add trailing newline.

File: launch/urg_node_launch.py
Comment (clalancette) [inline launch/urg_node_launch.py]: This looks like old-style launch file, suggest updating this to the new-style in Dashing and Eloquent.

File: src/urg_node_driver.cpp
Comment (clalancette) [inline src/urg_node_driver.cpp]: Is this line necessary?

File: src/urg_node_driver.cpp
Comment (clalancette) [inline src/urg_node_driver.cpp]:
```suggestion
    } catch (const std::runtime_error & e) {
```

**Comment by Karsten1987** [comment]
Thanks for the review. That PR and related ones need some more love before being merged. I hope to address this over the upcoming holidays.

**Comment by Paul-Ajay** [comment]
@Karsten1987 PR created https://github.com/bosch-robotics-na/urg_node/pull/1

**Comment by Paul-Ajay** [comment]
@Karsten1987 I just created another PR to fix the launch files and parameters. Feel free to review it. https://github.com/bosch-robotics-na/urg_node/pull/2

File: CMakeLists.txt
Comment (Karsten1987) [reply CMakeLists.txt]: boost is still being used for CRC. I don't have a simple STL solution for this. Hence, I am leaving it in.

File: src/urg_node_driver.cpp
Comment (Karsten1987) [reply src/urg_node_driver.cpp]: I believe this yields a warning of `unused variable` otherwise. I don't know too much about the original use case of this synchronization as for I just "commented it out".

File: include/urg_node/urg_node_driver.hpp
Comment (Karsten1987) [reply include/urg_node/urg_node_driver.hpp]: I'd like to postpone this for a follow PR. I'd like to reformat the nodes later on to be a lifecycle node and then in the same stroke make them composable.

**Comment by Karsten1987** [comment]
This is up for another round of review. I've enabled the ROS2 default linters and applied a bunch of suggestions from the first round of review.

**Comment by Paul-Ajay** [comment]
@Karsten1987 I am getting a compilation error after pulling latest changes in this PR 
```
Could not find a package configuration file provided by "boost" with any of
  the following names:
    boostConfig.cmake
    boost-config.cmake
```

**Comment by Karsten1987** [comment]
Let's see:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8944/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4713/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7314/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8850/)

File: CMakeLists.txt
Comment (clalancette) [inline CMakeLists.txt]: This wouldn't link for me; I had to remove the `SHARED` from this line.  I suspect it's because the `urg_c_wrapper` library above is not marked SHARED, and they need to agree to be linked together.  Suggest removing `SHARED` from this line.

File: CMakeLists.txt
Comment (clalancette) [inline CMakeLists.txt:23]: This is kind of weird, since we don't use `Boost::thread` anywhere.  But there doesn't seem to be a component for `crc`; it just sort of comes in as part of Boost.  That deserves at least a comment until we can remove this lingering boost dependency.

File: launch/urg_node_ethernet.yaml
Comment (clalancette) [inline launch/urg_node_ethernet.yaml]: Nit: add a trailing newline.

File: urg.repos
Comment (clalancette) [inline urg.repos]: It seems to me like `diagnostics`, `ros2_console`, and `ros2_time` are no longer required to compile the urg stuff.  At least, I didn't need them to compile.  If that's the case, please remove the from this repos file.

File: src/urg_node.cpp
Comment (clalancette) [inline src/urg_node.cpp:570]: `rclcpp::shutdown()` at the end is common.

File: src/urg_c_wrapper.cpp
Comment (clalancette) [inline src/urg_c_wrapper.cpp:59]: This doesn't seem to be necessary since it is a class variable.  Same below.

File: src/urg_node_driver.cpp
Comment (clalancette) [reply src/urg_node_driver.cpp]: I don't think it does cause a warning.  Because this is now a class variable, the compiler will no longer generate the `unused variable` warning (I confirmed this locally).  So I'd just remove this.

File: src/urg_node_driver.cpp
Comment (clalancette) [inline src/urg_node_driver.cpp]: This feels like a layering violation.  Particularly in the case where this is a composable node in the future, we don't necessarily want to shutdown the whole world, we just want to quit ourselves.  So maybe it makes more sense to just re-raise the exception?

File: src/urg_node_driver.cpp
Comment (clalancette) [inline src/urg_node_driver.cpp]: This is another case that feels like a layering violation, and we might just want to re-raise the error.

File: src/urg_node_driver.cpp
Comment (clalancette) [inline src/urg_node_driver.cpp]:
```suggestion
  } catch (const std::runtime_error & e) {
```

File: src/urg_node_driver.cpp
Comment (clalancette) [inline src/urg_node_driver.cpp]:
```suggestion
  } catch (const std::runtime_error & e) {
```

File: src/urg_node_driver.cpp
Comment (clalancette) [inline src/urg_node_driver.cpp]:
```suggestion
  } catch (const std::exception & e) {
```

**Review by clalancette — COMMENTED** [review]
I left a bunch more comments.  I think this is getting close to ready.

File: CMakeLists.txt
Comment (Karsten1987) [reply CMakeLists.txt]: that sounds strange. Why wouldn't you be able to link a static library to a shared one?

File: README.md
Comment (Karsten1987) [inline README.md]: with the set of PRs open (and ideally released) these instructions should be not necessary in the first place.

Comment (clalancette) [reply README.md]: Sure, makes sense to me.  Let's just remove this bit then.

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt]: Probably because the shared one is built with `-fPIC` and the static one without it.  Does it build for you?

Comment (Karsten1987) [reply CMakeLists.txt]: yup, builds cleanly for me - must be because of OSX being the better OS ;-)

File: src/urg_c_wrapper.cpp
Comment (Karsten1987) [reply src/urg_c_wrapper.cpp:59]: it somehow is. Otherwise my compiler complains about an unused class variable:
```
src/urg_node/include/urg_node/urg_c_wrapper.hpp:255:16: error: private field 'adj_alpha_' is not used [-Werror,-Wunused-private-field]
  const double adj_alpha_ = .01;
```

Now, I guess we could also just delete it all together, but as I am not familiar with what this was intended to do I opted for simply commenting it out.

File: src/urg_node_driver.cpp
Comment (Karsten1987) [reply src/urg_node_driver.cpp]: done

File: src/urg_node_driver.cpp
Comment (Karsten1987) [reply src/urg_node_driver.cpp]: done in 32fef1a

File: CMakeLists.txt
Comment (Karsten1987) [reply CMakeLists.txt]: Re-iterating over this, I didn't see the point of having the two files compiled into its individual libraries, so I've put them together.
In order to make them components, they have to be a shared lib in order for poco to load that thing during runtime. Please let me know if that works for you.

Comment (clalancette) [reply CMakeLists.txt]: Sort of.  I now had to add the SHARED keyword to both the laser_proc and the urg_c stuff; with all of that in place, it now works.  I'll open additional PRs to those two to show you what I mean.

**Review by clalancette — COMMENTED** [review]
This is overall looking pretty good.  One thing I'll suggest is to update .travis.yml so that PR testing starts succeeding.

**Comment by Karsten1987** [comment]
The current PR doesn't have any travis.yml anymore. I believe it's a legacy github thing that the job is still being run for this PR. I think it's good to have a github actions PR following this one.

**Comment by Karsten1987** [comment]
Let's see again:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9006/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4773/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7375/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8941/)

**Comment by Karsten1987** [comment]
Windows again: [Build Status](https://ci.ros2.org/job/ci_windows/8942/)

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
@chadrockey @tonybaltovski  This looks good to me now, and is passing CI.  I think this is ready to merge.  Please give it a look and let us know of any comments you have.  If I don't hear anything by the end of the week, I'll go ahead and merge and we can follow-up later.

**Comment by Karsten1987** [comment]
For completeness, a successful round of CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9008/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4775/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7376/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8948/) // Windows to be expected unstable as we issue a `non-supported` cmake warning for `urg_c` and `urg_node`

**Comment by clalancette** [comment]
@Karsten1987 Sorry to say this, but this branch does still have a `.travis.yml`, which is why Travis is failing :).

I think it makes sense to spend some time and port the travis changes over to work with ROS 2.  It's not too hard, and at least gives a baseline level of "do things work on Linux".  You can see an example of what I've done elsewhere in https://github.com/ros-drivers/phidgets_drivers/blob/e87cc88829160f074ac378e7f85be4df75282be1/.travis.yml and https://github.com/ros-drivers/phidgets_drivers/blob/e87cc88829160f074ac378e7f85be4df75282be1/Dockerfile-dashing .

File: .travis.yml
Comment (mikaelarguedas) [inline .travis.yml]: For all jobs to build these packages from source, this should be put in the `install` step. Currently only the `nightly` job builds the dependencies from source explaining why the other jobs fail to resolve `urg_c` rosdep key.

Comment (Karsten1987) [reply .travis.yml]: Let's see if that works with https://github.com/ros-drivers/urg_node/pull/50/commits/1882fe6ce8ae9db24b9e6e534656f533939694d2

thanks for the quick response!

**Comment by Karsten1987** [comment]
@clalancette the travis file are updated and work as expected now. They still fail though as the officially supported docker images from the OSRF don't install boost which is one dependency of this package. Please advise how to proceed further.

**Comment by mikaelarguedas** [comment]
> boost which is one dependency of this package

Listing boost as a dependency in the package.xml would make sense then and likely solve your problem

**Comment by Karsten1987** [comment]
thanks @mikaelarguedas 
Dashing and Eloquent are passing now, the nightlies are failing as diagnostics has not been released into Foxy yet.

I think this travis file has to be rewritten once the urg related dependencies are released into Eloquent

**Comment by clalancette** [comment]
> Dashing and Eloquent are passing now, the nightlies are failing as diagnostics has not been released into Foxy yet.

I'd suggest getting rid of the nightlies.  They are basically always going to fail, except for the short time window right before a release where the package is released but the whole distro hasn't been frozen.  Then this will be green too.

> 
> I think this travis file has to be rewritten once the urg related dependencies are released into Eloquent

Yeah, makes sense.  We'll do a follow-up once that is done.

**Comment by mikaelarguedas** [comment]
>  the nightlies are failing as diagnostics has not been released

What about providing a different `additional_repos.repos` file to the nightly job that includes the unreleased dependencies ?

**Comment by Karsten1987** [comment]
If all the related packages are hopefully eventually released, I think we can also just rely on the PR jobs.

**Comment by clalancette** [comment]
This is good for me now.  We can follow-up with additional checking later.  I'm going to go ahead and merge it.

**Pull request merged**
